launchers/junit/src/testFixtures/java/org/eclipse/dataspaceconnector/junit/launcher/EdcRuntimeExtension.java

```diff
@@ -0,0 +1,184 @@
+        // Find the project root directory, moving up the directory tree
+        var root = findRoot(new File(".").getCanonicalFile());
+        if (root == null) {
+            throw new EdcException("Could not find " + GRADLE_WRAPPER + " in parent directories.");
+        }
+
+        // Run a Gradle custom task to determine the runtime classpath of the module to run
+        String[] command = {
+                new File(root, GRADLE_WRAPPER).getCanonicalPath(),
+                "-q",
+                moduleName + ":printClasspath"
+        };
+        Process exec = Runtime.getRuntime().exec(command);
+        InputStream inputStream = exec.getInputStream();
+        var classpathString = new String(inputStream.readAllBytes());
+        if (exec.waitFor() != 0) {
+            throw new EdcException("Failed to run gradle command: " + String.join(" ", command));
+        }
+
+        // Replace subproject JAR entries with subproject build directories in classpath.
+        // This ensures modified classes are picked up without needing to rebuild dependent JARs.
+        var classPathEntries = Arrays.stream(classpathString.split(":|\\s"))
+                .filter(s -> !s.isBlank())
+                .flatMap(p -> resolveClassPathEntry(root, p))
+                .toArray(URL[]::new);
+
+        // Create a ClassLoader that only has the target module class path, and is not
+        // parented with the current ClassLoader.
+        var classLoader = URLClassLoader.newInstance(classPathEntries,
+                ClassLoader.getSystemClassLoader());
+
+        // Temporarily inject system properties.
+        var savedProperties = (Properties) System.getProperties().clone();
+        properties.forEach(System::setProperty);
+
+        var latch = new CountDownLatch(1);
+
+        runtimeThread = new Thread(() -> {
+            try {
+
+                // Make the ClassLoader available to the ServiceLoader.
+                // This ensures the target module's extensions are discovered and loaded at runtime boot.
+                Thread.currentThread().setContextClassLoader(classLoader);
+
+                // Boot EDC runtime.
+                super.beforeTestExecution(extensionContext);
+
+                latch.countDown();
+            } catch (Exception e) {
+                throw new EdcException(e);
+            }
+        });
+
+        MONITOR.info("Starting module " + moduleName);
+        // Start thread and wait for EDC to start up.
+        runtimeThread.start();
+
+        if (!latch.await(10, SECONDS)) {
+            throw new EdcException("Failed to start EDC runtime");
+        }
+
+        MONITOR.info("Module " + moduleName + " started");
+        // Restore system properties.
+        System.setProperties(savedProperties);
+    }
+
+
+    @Override
+    public void afterTestExecution(ExtensionContext context) throws Exception {
+        runtimeThread.join();
```

Comment:
In the cases of errors before `runtimeThread` instantiation this throw an NPE, better wrap it with a null check

Reply:
Great point, implemented